abci/vote_extensions.go

```diff
@@ -44,7 +65,45 @@ func NewVoteExtensionHandler(mp VoteExtensionMempool, txDecoder sdk.TxDecoder,
 // returns it in its vote extension.
 func (h *VoteExtensionHandler) ExtendVoteHandler() ExtendVoteHandler {
 	return func(ctx sdk.Context, req *RequestExtendVote) (*ResponseExtendVote, error) {
-		panic("implement me")
+		var voteExtension []byte
+
+		// Reset the cache if necessary
+		h.checkStaleCache(ctx.BlockHeight())
+
+		// Iterate through auction bids until we find a valid one
+		auctionIterator := h.mempool.AuctionBidSelect(ctx)
+		txsToRemove := make(map[sdk.Tx]struct{})
+
+		for ; auctionIterator != nil; auctionIterator = auctionIterator.Next() {
+			bidTx := auctionIterator.Tx()
+
+			// Verify the bid tx can be encoded and included in vote extension
+			bidBz, err := h.txEncoder(bidTx)
+			if err != nil {
+				txsToRemove[bidTx] = struct{}{}
+				continue
+			}
+
+			hashBz := sha256.Sum256(bidBz)
+			hash := hex.EncodeToString(hashBz[:])
+
+			// Validate the auction transaction and cache result
+			if err := h.verifyAuctionTx(ctx, bidTx); err != nil {
+				h.cache[hash] = err
+				txsToRemove[bidTx] = struct{}{}
+			} else {
+				h.cache[hash] = nil
+				voteExtension = bidBz
+				break
+			}
+		}
+
+		// Remove all invalid auction bids from the mempool
+		for tx := range txsToRemove {
+			h.RemoveTx(tx)
+		}
```

Comment:
Do we want to be removing invalid bids in the vote extension handlers?

Reply:
nit: I think vote extension handlers should be idempotent and simple as possible. I.e. let's not remove or modify the mempool in `ExtendVote`.